src/shaders/MaterialX/color4.h

```diff
@@ -15,312 +15,267 @@ struct color4
+    return color4 (smoothstep(low.rgb, high.rgb, in.rgb),
+                   smoothstep(low.a, high.a, in.a)
+                   );
 }
 
 color4 smoothstep(float low, float high, color4 in)
 {
-    color4 out = in;
-    out.rgb = smoothstep(low, high, in.rgb);
-    out.a = smoothstep(low, high, in.a);
-    return out;
+    return color4 (smoothstep(low, high, in.rgb),
+                   smoothstep(low, high, in.a)
+                   );
 }
 
 color4 remap(color4 in, color4 inLow, color4 inHigh, color4 outLow, color4 outHigh, int doClamp)
 {
       //remap in from [inLow, inHigh] to [outLow, outHigh], optionally clamping to the new range
 
-      color4 out = in;
-      out.rgb = remap(in.rgb, inLow.rgb, inHigh.rgb, outLow.rgb, outHigh.rgb, doClamp);
-      out.a = remap(in.a, inLow.a, inHigh.a, outLow.a, outHigh.a, doClamp);
-      return out;
+      return color4 (remap(in.rgb, inLow.rgb, inHigh.rgb, outLow.rgb, outHigh.rgb, doClamp),
+                     remap(in.a, inLow.a, inHigh.a, outLow.a, outHigh.a, doClamp)
+                     );
 }
 
 color4 remap(color4 in, float inLow, float inHigh, float outLow, float outHigh, int doClamp)
 {
       //remap in from [inLow, inHigh] to [outLow, outHigh], optionally clamping to the new range
-      color4 out = in;
-      out.rgb = remap(in.rgb, inLow, inHigh, outLow, outHigh, doClamp);
-      out.a = remap(in.a, inLow, inHigh, outLow, outHigh, doClamp);
-      return out;
+      return color4 (remap(in.rgb, inLow, inHigh, outLow, outHigh, doClamp),
+                     remap(in.a, inLow, inHigh, outLow, outHigh, doClamp)
+                     );
 }
 
 color4 fgamma(color4 in, color4 g)
 {
-    color4 out = in;
-    out.rgb = fgamma(in.rgb, g.rgb);
-    out.a = fgamma(in.a, g.a);
-    return out;
+    return color4 (fgamma(in.rgb, g.rgb),
+                   fgamma(in.a, g.a)
+                   );
 }
 
 color4 fgamma(color4 in, float g)
 {
-    color4 out = in;
-    out.rgb = fgamma(in.rgb, g);
-    out.a = fgamma(in.a, g);
-    return out;
+    return color4 (fgamma(in.rgb, g),
+                   fgamma(in.a, g)
+                   );
 }
 
 color4 invert(color4 in, color4 amount)
 {
-    color4 out = in;
-    out.rgb = amount.rgb - in.rgb;
-    out.a = amount.a - in.a;
-    return out;
+    return color4 (amount.rgb - in.rgb,
+                   amount.a - in.a
+                   );
 }
 
 color4 invert(color4 in, float amount)
 {
-    color4 out = in;
-    out.rgb = color(amount) - in.rgb;
-    out.a = amount - in.a;
-    return out;
+    return color4 (color(amount) - in.rgb,
+                   amount - in.a
+                   );
 }
 
 color4 invert(color4 in)
 {
-    color4 out = invert(in, 1.0);
-    return out;
+    return invert(in, 1.0);
 }
 
 color4 clamp(color4 in, color4 low, color4 high)
 {
-    color4 out = in;
-    out.rgb = clamp(in.rgb, low.rgb, high.rgb);
-    out.a = clamp(in.a, low.a, high.a);
-    return out;
+    return color4 (clamp(in.rgb, low.rgb, high.rgb),
```

Comment:
Silly question (maybe for Doug): for an RGB+alpha, is the right definition of "invert" to also invert the alpha? When would that be helpful? Or should alpha be left alone and only the color inverted? Er, or is the right thing that invert(rgba) == (a-rgba, a), in other words "invert" only with respect to the total coverage?The same question might be asked of other color4 functions.

The unobviousness of exactly what the math should mean on a color+alpha is a big part of why I didn't make a color4 as a fundamental type in the language. :-)